### PR TITLE
Notification Module

### PR DESCRIPTION
**Pull request by adhyans**
This module will allow developers to integrate notifications with their functionality. Here are some wikis on which you can read more about this module.

1) [Notification Module Design](https://mifosforge.jira.com/wiki/pages/viewpage.action?pageId=133070890)
2) [Notification API Developer User Guide](https://mifosforge.jira.com/wiki/display/MDZ/Notification+API+Developer+User+Guide)
3) [Configure Notifications-End User guide](https://mifosforge.jira.com/wiki/display/docs/Configure+Notifications)

### CODE REVIEW

File: fineract-provider/dev-dependencies.gradle [hunk 1]
```diff
@@ -90,6 +90,10 @@ dependencies {
                 [group: 'org.apache.httpcomponents', name: 'httpclient', version: '4.3.5'],
                 // Once we've switched to Java 8 this dep can be removed.
                 //[group: 'com.google.code.findbugs', name: 'jsr305', version: '3.0.0']
+                [group: 'org.springframework', name:'spring-jms'],
+                [group: 'org.apache.activemq', name: 'activemq-broker'],
+                [group: 'com.mockrunner', name: 'mockrunner-jms', version: '1.0.6'],
```
Comment (nazeer1100126) [inline fineract-provider/dev-dependencies.gradle:95]: Are these dependencies should be part of testCompile?

File: fineract-provider/dependencies.gradle [hunk 2]
```diff
@@ -91,6 +91,11 @@ dependencies {
                 [group: 'org.apache.httpcomponents', name: 'httpclient', version: '4.3.5'],
                 // Once we've switched to Java 8 this dep can be removed.
                 //[group: 'com.google.code.findbugs', name: 'jsr305', version: '3.0.0']
+                [group: 'org.springframework', name:'spring-jms'],
+                [group: 'org.apache.activemq', name: 'activemq-broker'],
+                [group: 'com.mockrunner', name: 'mockrunner-jms', version: '1.0.6'],
```
Comment (nazeer1100126) [inline fineract-provider/dependencies.gradle:96]: Are these dependencies should be part of testCompile?

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/NotificationApiTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        Utils.initializeRESTAssured();
+        this.requestSpec = new RequestSpecBuilder().setContentType(ContentType.JSON).build();
+        this.requestSpec.header("Authorization", "Basic " + Utils.loginIntoServerAndGetBase64EncodedAuthenticationKey());
+        this.responseSpec = new ResponseSpecBuilder().expectStatusCode(200).build();
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testNotificationRetrieval() {
+        HashMap<String, Object> response = (HashMap<String, Object>) NotificationHelper.getNotifications(this.requestSpec,
```
Comment (nazeer1100126) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/NotificationApiTest.java:52]: Where is the actual test cases for notifications? You are just verifying the response?
Or is there any limitation here?

File: fineract-provider/src/main/java/org/apache/fineract/notification/data/NotificationData.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+    public String getTenantIdentifier() {
+        return tenantIdentifier;
+    }
+
+    public void setTenantIdentifier(String tenantIdentifier) {
+        this.tenantIdentifier = tenantIdentifier;
+    }
+
+    public static class NotificationBuilder {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/notification/data/NotificationData.java:152]: You need move this NotificationBuilder to separate java file

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationGenerator.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                ", objectIdentifier=" + objectIdentifier +
+                ", action='" + action + '\'' +
+                ", actor='" + actor + '\'' +
+                ", isSystemGenerated=" + isSystemGenerated +
+                ", notificationContent='" + notificationContent + '\'' +
+                ", createdAt=" + createdAt +
+                '}';
+    }
+
+    public static class NotificationGeneratorBuilder {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationGenerator.java:127]: NotificationGeneratorBuilder should be separate java file as NotificationGenerator should follow SRP? NotificationGenerator name will confuse the users as it is not a generator but it is Entity.

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationMapper.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+    @ManyToOne
+    @JoinColumn(name = "user_id")
+    private AppUser userId;
+
+    @Column(name = "is_read")
+    private boolean isRead;
+
+    @Column(name = "created_at")
+    private String createdAt;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationMapper.java:43]: createdAt should be a date object?

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationMapper.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    public boolean isRead() {
+        return isRead;
+    }
+
+    public void setRead(boolean read) {
+        isRead = read;
+    }
+
+    public static class NotificationMapperBuilder {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationMapper.java:89]: NotificationMapperBuilder should be a separate java file to follow SRP?

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationGenerator.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    private String objectType;
+
+    @Column(name = "object_identifier")
+    private Long objectIdentifier;
+
+    @Column(name = "action")
+    private String action;
+
+    @Column(name = "actor")
+    private String actor;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationGenerator.java:41]: Actor should be an id as two persons can have same name?

File: fineract-provider/src/main/java/org/apache/fineract/notification/eventandlistener/NotificationEventListener.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                    .loadTenantById(notificationData.getTenantIdentifier(), false);
+            ThreadLocalContextUtil.setTenant(tenant);
+
+            AppUser currentAppUser = appUserRepository.findAppUserByName(notificationData.getActor());
+
+            List<Long> userIds;
+            if (notificationData.getUserId() == null) {
+                userIds = notificationData.getUserIds();
+            } else {
+                userIds = new ArrayList<>();
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/notification/eventandlistener/NotificationEventListener.java:73]: Is there any chance that userId is null but userIds is not null in notificationData?

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/NotificationApiTest.java (hunk 3 above)
Comment (adhyans) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/NotificationApiTest.java:52]: I have included the test for sending and receiving of notifications in `test/java/org/apache/fineract/notification`

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationMapper.java (hunk 6 above)
Comment (adhyans) [reply fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationMapper.java:43]: I have stored a formatted string of date object in the database i.e why I used String here.

File: fineract-provider/src/main/java/org/apache/fineract/notification/data/NotificationData.java (hunk 4 above)
Comment (adhyans) [reply fineract-provider/src/main/java/org/apache/fineract/notification/data/NotificationData.java:152]: The use of this static class is only in `NotificationData` class. I have used this static inner class to implement a builder pattern. Is it advisable to move to another file ?

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationGenerator.java (hunk 5 above)
Comment (adhyans) [reply fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationGenerator.java:127]: The use of this static class is only in `NotificationGenerator` class. I have used this static inner class to implement a builder pattern. Is it advisable to move to another file?

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationMapper.java (hunk 7 above)
Comment (adhyans) [reply fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationMapper.java:89]: The use of this static class is only in `NotificationMapper` class. I have used this static inner class to implement a builder pattern. Is it advisable to move to another file?

File: fineract-provider/src/main/java/org/apache/fineract/notification/eventandlistener/NotificationEventListener.java (hunk 9 above)
Comment (adhyans) [reply fineract-provider/src/main/java/org/apache/fineract/notification/eventandlistener/NotificationEventListener.java:73]: I have used two different variables, one for when the notification is sent to only one user and another variable when notification is sent to more than one users. I will refactor it to use only one variable :)

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationGenerator.java (hunk 8 above)
Comment (adhyans) [reply fineract-provider/src/main/java/org/apache/fineract/notification/domain/NotificationGenerator.java:41]: Refactored actor  to an id :)

File: fineract-provider/src/main/java/org/apache/fineract/notification/data/NotificationData.java (hunk 4 above)
Comment (nazeer1100126) [reply fineract-provider/src/main/java/org/apache/fineract/notification/data/NotificationData.java:152]: Is it good to have two public class definitions in same java file?

**Pull request closed without merging**